### DISCUSSION TITLE
Still getting Binder Error: column xxxx must appear in the GROUP BY clause or be used in an aggregate function

### DISCUSSION THREAD

**Discussion by sahoosatish** [duckdb/duckdb#4149] [General]
Hi, I am using duckdb as the backend in python to use SQL on pandas. I get the following error with the following query-
Version-I am using duckdb 0.4.0

```sql
select ,min(tarfyear) as tarfyearmin from
(select a.,b.fyear as tarfyear from
comp a left join comp b
on a.gvkey=b.gvkey and a.fyear<=b.fyear and b.impd=1)
group by gvkey,fyear having tarfyear=min(tarfyear)
```

I get the error "Binder Error: column unnamed_subquery7.tarfyear must appear in the GROUP BY clause or be used in an aggregate function"
I have tried using both having 'tarfyear=min(tarfyear)' and 'having tarfyear=tarfyearmin'
Both give the same error.

Please advise.
Thanks

**Comment by Alex-Monahan** [comment]
Howdy! You'll want back tics for code highlighting, not apostrophes! :-)

How about using:
```sql
having tarfyearmin=min(tarfyear)
```
The column in a having clause needs to either be an aggregate function or exist after the group by is complete, and tarfyear no longer exists post grouping.

Could you describe what you expect to get a bit more clearly? I'm not able to understand the intent of your query.

Thanks!

**Reply by sahoosatish** [reply]
Hi Alex. Thanks for the quick response. And sorry for the code formatting.
Regarding using 'having tarfyearmin=min(tarfyear)' as i mentioned I tried that and it fails with the same error. Though the other thread mentions what you are suggesting as the solution for this issue, it doesn't seem to be working in this case.

Regarding the logic of the code-
I am trying to identify the first fyear after the current fyear the impd indicator becomes 1 for every gvkey,fyear pair. Hope it makes sense.

Thanks

**Reply by Alex-Monahan** [reply]
My suggestion is slightly different than the other options that you mentioned trying. Did you test it out?

Could you provide a reproducible test case that includes either your current dataset, or an example dataset with the same problem?

Thanks!

**Comment by sahoosatish** [comment]
Hi Alex. Thanks for your suggestion. I did try that but it doesn't work. I played around a little bit more by breaking the query down, but I guess I need to step back a bit here. I think since I am used to running SQL in SAS I am struggling here. Which version of SQL does duckdb  implement ? I see that the following simple query is also not allowed in duck db. I think many SQL versions accept this query. Please let me know if I am missing something here. Or is this the expected behavior in duck db ? Thanks
```sql

select *,min(sale) as salemin
from comp
group by gvkey,year

```
RuntimeError: Binder Error: column "tic" must appear in the GROUP BY clause or be used in an aggregate function
tic is essentially the first column in the dataset after gvkey.

**Reply by Mytherin** [reply]
This is expected behavior as defined by the SQL standard, and almost all SQL systems (minus SQLite) will produce an error for this query. If you are doing an aggregate you are folding the table along certain axes (the `GROUP BY` columns), and you either need to specify how to do this folding (provide an aggregate function) or fold along that axis (add it to the list of `GROUP BY` columns).

You could add a `FIRST` aggregate over the columns that are not part of the `GROUP BY`, which is what SQLite would do for you by default.
